proposals/partial-type-inference-v2.md

```diff
@@ -0,0 +1,389 @@
+    ...
+    F<IList<_>>(new List<int>());
+    ```
+
+  * Consider an example above. 
+    We have now two type variables `T1` and `_`. From the first bound, we get that `IList<_>` is a shape bound of `T1`(Ignore now the type of bound, it would be the same in other types of bound).
+    When we investigate the second bound `List<int>`, we will figure out that it would be a lower bound of `T1`.
+    But now, we have to somehow propagate the int type to the `_` type variable, because it relates to it.
+    That means, in the process of adding new bounds, we have to also propagate this info through bounds, which contain unfixed type variables.
+    In this case, we do additional inference of `IList<_>` and `List<int>` yielding exact bound `int` of `_`.
+
+* Explanation of type-dependence
+  * Type-dependence is required because, till this time when a type variable had any bounds, it didn't contain any unfixed type variable.
+    It was important because we could do the parameter fixation, where we work with exact types(not unfixed type variables).
+    However now, the type variable can contain bounds containing unfixed type variables.
+    We have to ensure that we will not start fixing the type variable till these unfixed type variables are unfixed(In some cases, we can be in a situation, where this dependency will form a cycle. In this case, we will allow the fixation earlier).
+  
+  * We use the previous example.
+    After the first phase. `T1` has bounds `IList<_>` and `List<int>`. `_` has bound `int`.
+    In this situation, we can't start to fix `T1` because `_` is not fixed yet.
+    `T1` is type-dependent on `_`.
+    So, we will first fix `_`, which becomes `int`.
+    Then, `T1` is not type-dependent anymore, because all bounds don't contain any unfixed type variables.
+    `IList<_>` is now `IList<int>` after the `_` fixation.
+    We can fix `T1` now.
+
+* Explanation of shape-dependence
+  * A similar thing is for shape-dependence.
+    Although it cares about bounds received from the type argument list.
+    We want a shape bound to be exact (not containing any unfixed type variables) because it is later important for the fixation.
+    An intention is to keep the exact form of the given hint(`IList<_>`).
+
+  * Example: Given `IList<_>` as a type argument, when we treat nullability, we want the hinted type parameter to be non-nullable(not `IList<_>?`).
+    It can happen, other bounds would infer the nullable version, and although `IList<_>` can be converted to `IList<_>?`, it is not the user's intention.
+
+### Compile-time checking of dynamic member invocation
+
+We change the [compile-time checking](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1265-compile-time-checking-of-dynamic-member-invocation).
+
+- First, if `F` is a generic method and type arguments were provided, then if the method group is *partial_inferred*, an binding error occurs since it is not supported in the runtime binding.
```

Comment:
We will likely decide to perform a partial inference at compile time when a single candidate is found for the dynamic invocation. It's not urgent to revise this proposal to get this precisely right. Certain details of this may be subject to change. See https://github.com/dotnet/csharplang/blob/main/proposals/params-collections.md#dynamic-vs-static-binding.